graphbolt/src/feature_cache.h

```diff
@@ -0,0 +1,77 @@
+/**
+ *   Copyright (c) 2023, GT-TDAlab (Muhammed Fatih Balin & Umit V. Catalyurek)
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ * @file feature_cache.h
+ * @brief Feature cache implementation on the CPU.
+ */
+#ifndef GRAPHBOLT_FEATURE_CACHE_H_
+#define GRAPHBOLT_FEATURE_CACHE_H_
+
+#include <torch/custom_class.h>
+#include <torch/torch.h>
+
+#include <vector>
+
+namespace graphbolt {
+namespace storage {
+
+struct FeatureCache : public torch::CustomClassHolder {
+  /**
+   * @brief Constructor for the FeatureCache struct.
+   *
+   * @param shape The shape of the cache.
+   * @param dtype The dtype of elements stored in the cache.
+   */
+  FeatureCache(const std::vector<int64_t>& shape, torch::ScalarType dtype);
+
+  FeatureCache() = default;
+
+  /**
+   * @brief The cache query function. Allocates an empty tensor `values` and
+   * runs values[indices] = cache_tensor[positions] before returning it.
+   *
+   * @param positions The positions of the queries items.
+   * @param indices The indices of the queried items among the original keys.
+   * @param size The size of the original keys, hence the first dimension of
+   * the output shape.
+   * @param pin_memory Whether to pin the memory of the output values tensor.
+   *
+   * @return The values tensor is returned. Its memory is pinned if pin_memory
+   * is true.
+   */
+  torch::Tensor Query(
```

Comment:
The documentation should clarify the indices.size() might be smaller than size, as result only corresponding rows in the return Tensor have meaningful value.